### PR TITLE
Add additional features to IRC input

### PR DESCRIPTION
**Pull request by coolacid**
Original PR: https://github.com/elasticsearch/logstash/pull/1800

This PR adds the ability to collect more details (JOINs and PARTs) and add's the optional ability to request a NAMES list on X interval, thus giving statistics on number of users in channels.

### CODE REVIEW

File: lib/logstash/inputs/irc.rb [hunk 1]
```diff
@@ -62,27 +73,81 @@ def register
...
     end
     loop do
       msg = @irc_queue.pop
-      if msg.user
+      handle_response(msg, output_queue)
+    end
+  end # def run
+
+  def handle_response (msg, output_queue)
+      if @get_stats and msg.command.to_s == "353"
```
Comment (suyograo) [inline lib/logstash/inputs/irc.rb]: create constants for these numbers?

File: lib/logstash/inputs/irc.rb [hunk 2]
```diff
@@ -62,27 +73,81 @@ def register
...
+        # Got a names list event
+        # Count the users returned in msg.params[3] split by " "
+        users = msg.params[3].split(" ")
+        @user_stats[msg.channel.to_s] = (@user_stats[msg.channel.to_s] || 0)  + users.length
+      end
+      if @get_stats and msg.command.to_s == "366"
```
Comment (suyograo) [inline lib/logstash/inputs/irc.rb]: same here, create constants?

File: lib/logstash/inputs/irc.rb [hunk 3]
```diff
@@ -62,27 +73,81 @@ def register
...
   public
   def run(output_queue)
-    Thread.new(@bot) do |bot|
-      bot.start
+    @bot_thread = Stud::Task.new(@bot) do |bot|
+        bot.start
+    end
+    if @get_stats
+      @request_names_thread = Stud::Task.new do
+#        Stud.interval(30) do # @stats_interval * 60) do  # @bot goes away using this.
```
Comment (suyograo) [inline lib/logstash/inputs/irc.rb]: remove dead code

**Comment by coolacid** [comment]
Thanks @suyograo - That should be better?

File: lib/logstash/inputs/irc.rb [hunk 4]
```diff
@@ -33,20 +33,31 @@ class LogStash::Inputs::Irc < LogStash::Inputs::Base
...
+  # "#logstash password".
   #
   config :channels, :validate => :array, :required => true
 
   public
   def register
     require "cinch"
+    @user_stats = Hash.new
     @irc_queue = Queue.new
+    @catch_all = true if  @get_stats
```
Comment (electrical) [inline lib/logstash/inputs/irc.rb:60]: the `@catch_all` var is not used anywhere?

File: lib/logstash/inputs/irc.rb [hunk 5]
```diff
@@ -62,27 +73,84 @@ def register
...
+      end
+      if @get_stats and msg.command.to_s == RPL_ENDOFNAMES
+        # Got an end of names event, now we can send the info down the pipe.
+        event = LogStash::Event.new()
+        decorate(event)
+        event["channel"] = msg.channel.to_s
+        event["users"] = @user_stats[msg.channel.to_s]
+        event["server"] = "#{@host}:#{@port}"
+        output_queue << event
+      end
```
Comment (ph) [inline lib/logstash/inputs/irc.rb:126]: We have some redundancy here, we are checking the boolean value of `@get_stats` in  two differents conditionals.

We can replace it with one:

``` ruby
if @get_stats
  if msg.command.to_s == RPL_NAMREPLY
  #...
  elsif msg.command.to_s == RPL_ENDOFNAMES
  #...
  end
end
```

**Pull request closed without merging**